dlhub_sdk/utils/validation.py

```diff
@@ -0,0 +1,239 @@
+    """Recursively validate each of the elements in li against db_entry["item_type"]
+
+    Args:
+        li (list): The list whose items need to have their type validated
+        db_entry (dict): The entry that stores type data for li
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        TypeError: If any of the items' types do not match
+    """
+    path.append(["list", None])
+    for i, item in enumerate(li):
+        path[-1][1] = i
+        validate(item, db_entry["item_type"], path)
+    path.pop()
+
+
+def _validate_tuple(tup: tuple, db_entries: list, path: list) -> None:
+    """Recursively validate each of the elements in tup against the corresponding entry in db_entries
+
+    Args:
+        tup (tuple): The tuple whose items need to have their type validated
+        db_entries (list): The list of dicts that store the metadata for each item in tup
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If the number of entries and tuple elements do not match
+        TypeError: If any of the items' types do not match
+    """
+    if len(tup) != len(db_entries):
+        raise _generate_err(ValueError, path, msg=f"dl.run expected tuple of length {len(db_entries)}, recieved tuple with length {len(tup)}"
+                                                  + "{loc}")
+
+    path.append(["tuple", None])
+    for i, (given, expected) in enumerate(zip(tup, db_entries)):
+        path[-1][1] = i
+        validate(given, expected, path)
+    path.pop()
+
+
+def _validate_ndarray(arr: ndarray, shape: list, db_entry: dict, path: list) -> None:
+    """Compare the shape of arr with shape and recursively validate each of the elements in arr against db_entry["item_type"]
+
+    Args:
+        arr (ndarray): The ndarray whose shape and items need to be validated
+        shape (list): The shape that arr is expected to have
+        db_entry (dict): The entry where type data may or may not be retrieved
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If arr.shape does not match shape
+        TypeError: If any of the items' types do not match
+    """
+    shape = tuple(int(x) if x != "None" else x for x in shape)
+    shape_err = _generate_err(ValueError, path, msg=f"dl.run given improper input: expected ndarray.shape = {shape}, received shape: {arr.shape}"
+                                                    + "{loc}")
+
+    if len(arr.shape) != len(shape):
+        raise shape_err
+    if not all(map(lambda t: t[0] == t[1] or t[1] == "None", zip(arr.shape, shape))):
+        raise shape_err
+
+    entry = db_entry.get("item_type")
+    if entry:
+        path.append(["ndarray", None])
+        for i, item in ndenumerate(arr):
```

Comment:
Looping over the whole array is unnecessary as Numpy stores a type along with an array, and this could be slow for large arrays. Instead, how about we use the type already built into ndarrays (`arr.dtype`).

Reply:
The edge case I could see for `arr.dtype` not giving you the right information is if type is `object`. I rarely encounter ndarrays of python objects, but they're not impossible. Maybe you could put in some logic where you loop over contents only if the type is an object.

Potentially helpful: https://github.com/DLHub-Argonne/dlhub_sdk/blob/master/dlhub_sdk/utils/types.py#L5

Reply:
In general, the functions are written to handle objects, not their types. Because ndarrays have uniform type it's true that I can make the function faster by only checking the first item. This also solves the problem you mentioned (I think). I would do this by calling arr.item(0), but there are some notes in the [documentation](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.item.html#:~:text=suitable%20Python%20scalar-,Notes,-When%20the%20data) that I do not understand fully.